### PR TITLE
Fixes for when tracks are replaced, updates to Lavalink config, and fixes for when non-sharded clients get disconnected.

### PR DESCRIPTION
**Pull request by jdcunningham**
This fixes https://github.com/Devoxin/SharpLink/issues/20

This also fixes when a non-sharded client disconnects, it would modify the players dictionary and throw an exception since we can't modify the dictionary while it is being iterated. This meant the player lock would never release and cause further issues elsewhere when we attempt to wait on the lock to be released.

Also changed some things that you'll see in my notes along with my commits. Mostly just making sure we're not making too many duplicate operations.

### CODE REVIEW

**Comment by jdcunningham** [comment]
Just made another commit that should update SharpLink to be inline with the newest version of Lavalink. 

I've tested connecting to voice channels in various ways, and discovered that UserVoiceStateUpdate doesn't get called if the client is already in the channel we're trying to connect to. If I've done this properly, now when players get disconnected they should be fine to reconnect, and not cause any issues with clients already being connected to a voice channel they're trying to connect to again.

I haven't gotten a chance to check whether client disconnects are handled correctly after this commit, but I'll be sure to edit this comment as soon as it happens, and / or push a new commit to correct any issues that I discover (it happens very often to me so I'd give me 24 hours to actually test this, my internet is a pile of shit) Apart from this, I'm 99.9% sure that my changes shouldn't introduce any new issues, and if it does, I apologize.

Now, this does mean that there isn't any need for us to store the session id anymore because it won't actually be updated if we're already in the channel. So I decided that using the VoiceChannel variable to get the current session id would be safer.

I realize I didn't need to replace some of the various lines of code that throw errors that are already escaped and wouldn't cause issues, but I googled that throwing errors is actually less performant than just handling what's causing the error properly. I'll remove whatever you like, take as much time as you need to look at this. Thanks!

**Comment by Devoxin** [comment]
Regarding your latest commit, this would break compatibility for anything <3.1

File: SharpLink/LavalinkManagerConfig.cs [hunk 1]
```diff
@@ -5,10 +5,8 @@ namespace SharpLink
     public class LavalinkManagerConfig
     {
         // Default config follows the application.yml.example in the Lavalink GitHub located at https://github.com/Frederikam/Lavalink/blob/master/LavalinkServer/application.yml.example
-        public string WebSocketHost = "127.0.0.1";
-        public ushort WebSocketPort = 80;
-        public string RESTHost = "127.0.0.1";
-        public ushort RESTPort = 2333;
+        public string ServerAddress = "127.0.0.1";
+        public ushort ServerPort = 2333;
```
Comment (NovusTheory) [inline SharpLink/LavalinkManagerConfig.cs]: In particular with this setup is you're now setting the host and port for both the rest and websocket system which *can* be changed in older versions of Lavalink. Breaking the support for older versions entirely. If they were to have different setups for the configuration.

File: SharpLink/LavalinkPlayer.cs [hunk 2]
```diff
@@ -205,7 +195,7 @@ internal async Task UpdateSessionAsync(SessionChange change, object changeData =
                         JObject data = new JObject();
                         data.Add("op", "voiceUpdate");
                         data.Add("guildId", voiceServer.Guild.Id.ToString());
-                        data.Add("sessionId", sessionId);
+                        data.Add("sessionId", (await VoiceChannel.Guild.GetCurrentUserAsync()).VoiceSessionId);
```
Comment (NovusTheory) [inline SharpLink/LavalinkPlayer.cs]: You mentione that your intention was to also make this hopefully work for reconnect problems where the bot is already in a channel. The problem here with that is this VoiceSessionId will never be populated from when I was testing this. I would have further testing done to see what can be done with that issue

File: SharpLink/LavalinkManagerConfig.cs (hunk 1 above)
Comment (jdcunningham) [reply SharpLink/LavalinkManagerConfig.cs]: I'll revert these changes, but I think there needs to be a notice in the readme that states that the latest versions of Lavalink don't support two different host / ports. Devs will have to update their configs either way if they upgrade, and yeah at least it won't break backwards compatibility.

Comment (jdcunningham) [reply SharpLink/LavalinkManagerConfig.cs]: On second thought, I think I'll make a compromise, I'll change the ports to be identical, but leave the support for backwards compatibility, so this should be okay? I think I'll still make a note in the readme that just clarifies why there's two different declarations for the server host and port.

File: SharpLink/LavalinkPlayer.cs (hunk 2 above)
Comment (jdcunningham) [reply SharpLink/LavalinkPlayer.cs]: I can't repro this but I'll take your word for it. Definitely doesn't work though when the bot loses connection, so I'll keep looking into this. I would like to know how you produced this though, so I can test it before pushing another commit.

Comment (jdcunningham) [reply SharpLink/LavalinkPlayer.cs]: LOL, I think I actually solved this. I reverted all the session id functions back, and I tried finding a better place to get the voice session id, and it turns out that UserVoiceStateUpdated is called, it just wasn't already being caught with the current method so I didn't know about it. 😅 Anyway, this *should* work now, and will only be called *hopefully* when the client reconnects and the session id isn't already set. I actually like this way better.

Comment (jdcunningham) [reply SharpLink/LavalinkPlayer.cs]: Also, I pushed the log message inside of the actual player / session id check because I didn't want it to spit out "VOICE_STATE_UPDATE (yada) Updated" and using Reconnected would've been incorrect if we weren't actually reconnecting! 😄

Comment (jdcunningham) [reply SharpLink/LavalinkPlayer.cs]: Okay, does the new code still do this?

**Comment by jdcunningham** [comment]
I've made changes that should revert / fix the concerns you both mentioned. It was completely my bad for not acknowledging the need for backwards compatibility, no point in breaking something that doesn't need to be broken. (I still updated the config and readme to reflect the up-to-date Lavalink application.yml, but at least now it'll support the older versions like you asked) I'm a lot more confident about my solution for getting the proper voice session id upon reconnecting now. For some reason, my internet is actually doing pretty good right now, and I've tried simulating a client disconnect in the past to no avail. Don't feel rushed to preview any of my code, because I really don't want this pushed until I know for sure that the original issue with client reconnects I found is taken care of. I'll update this post (instead of the other one) as soon as it happens to me again, hopefully soon.

**Comment by jdcunningham** [comment]
**tldr**
From my testing this small fix should allow clients that get disconnected, properly be able to reconnect, even allow devs to use the `Connected` event from the Discord.Net api to directly rejoin any players that were destroyed after a connection is dropped. (Thank you so much NovusTheory for mentioning the url for the Discord gateway, made it way easier to test for this, and hopefully actually fixed this way sooner than just waiting hours for my internet to bog down) From what I can tell, I'm finished with this pr and all the issues I originally discovered are fixed / and I think I've taken care of your concerns.

**Explanation if you're curious, and why I think that using session id in this way is safe...**
Alright, so I've done more testing over the last couple of days and I think I've taken care of the issue I found with players being disconnected while they're being created. Essentially the issue is here... https://github.com/Devoxin/SharpLink/blob/4d233d7c57a21fd313d36b5df3b9cdbfd2844434/SharpLink/LavalinkManager.cs#L86-L99
When `JoinAsync` is called, sometimes when `LavalinkPlayer.ConnectAsync` is called, it will actually invoke `UserVoiceStateUpdated` two separate times. Once for a disconnect and then a proper connect afterwards. The problem is that `UserVoiceStateUpdated` isn't invoked until way after the player has been created, after we try connecting and after we've stored the player into the dictionary, what happens is the player will get destroyed because it thinks we've disconnected, even though that's exactly what we wanted. Whether it gives us a new session id or makes sure we're in the voice channel after a disconnect from the Discord gateway. This can be resolved by just simply checking to make sure the session id inside of `LavalinkPlayer` is set so we only destroy the player once it's been properly initialized. If you're worried that this could lead to further problems, I think that it's definitely possible, but completely improbable, because a disconnect would require us to be in the voice channel or a voice channel beforehand, and that means that we would've had to initialize the connection to begin with. So it's actually safe (or at least it should be) to just check for session id and ignore the disconnect if it doesn't exist yet, because we're waiting on the connect and don't care about disconnecting when we haven't even connected yet. I think that something that ensures we've properly connected to the channel would be nice, but that's for another day. It should be perfectly safe to use session id in a way that directly corresponds to whether the player was properly initialized so no new values have to be introduced or some weird way to track players after we get disconnected from the Discord gateway.

**Also**
The reason I removed this...
https://github.com/Devoxin/SharpLink/blob/4d233d7c57a21fd313d36b5df3b9cdbfd2844434/SharpLink/LavalinkManager.cs#L384-L385
Is because from what I could tell, when we do `VoiceChannel.ConnectAsync` inside of `LavalinkPlayer`, it actually generates a new session id if we're already in the channel. I could be wrong, and if there's a reason for this code (I don't know who actually put it there, I don't know how to search commits to see that), I don't mind leaving it in, because from what I can tell it won't have any negative impact either way, whether or not it's in because `VoiceChannel.ConnectAsync` will just do whatever it wants to do, and I think we're handling the connections properly with `UserVoiceStateUpdated`.
